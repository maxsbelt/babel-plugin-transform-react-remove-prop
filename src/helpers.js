const IDENTIFIERS = ['createElement', 'h']
export const DEFAULT_ATTRIBUTE = 'data-test-id'

export const isReactCallExpression = (t, callee) => {
  if (
    t.isMemberExpression(callee) && (
      !callee.object || callee.object.name !== 'React'
      || !callee.property || callee.property.name !== 'createElement'
    )
  ) return false
  if (t.isIdentifier(callee) && !IDENTIFIERS.includes(callee.name)) return false
  return true
}

export const isAssignExpression = (t, arg) => t.isCallExpression(arg)
  && arg.callee
  && arg.callee.property
  && arg.callee.property.name === 'assign'

const createFindIndex = (t, state) => {
  const attrs = state.opts && Array.isArray(state.opts.attrs)
    ? state.opts.attrs
    : [DEFAULT_ATTRIBUTE]
  return (item) => {
    if (t.isJSXAttribute(item)) return item.name && attrs.includes(item.name.name)
    return item.key && attrs.includes(item.key.value)
  }
}

const getArray = (t, arg) => {
  if (t.isObjectExpression(arg)) return arg.properties
  if (t.isJSXOpeningElement(arg)) return arg.attributes
  if (t.isJSXSpreadAttribute(arg) && arg.argument && Array.isArray(arg.argument.properties)) {
    return arg.argument.properties
  }
  return []
}

export const getCreateRemove = (t) => (state) => {
  const findIndex = createFindIndex(t, state)
  return (arg) => {
    const array = getArray(t, arg)
    let index = -1
    do {
      index = array.findIndex(findIndex)
      if (index !== -1) array.splice(index, 1)
    } while (index !== -1)
  }
}
