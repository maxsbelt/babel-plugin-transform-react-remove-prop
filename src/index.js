import * as helpers from './helpers'

export default ({ assertVersion, types }) => {
  assertVersion(7)
  const createRemove = helpers.getCreateRemove(types)
  return {
    name: 'transform-react-remove-prop',
    visitor: {
      JSXOpeningElement(path, state) {
        const remove = createRemove(state)
        path.container.openingElement.attributes.forEach(remove)
        remove(path.container.openingElement)
      },
      CallExpression(path, state) {
        if (!helpers.isReactCallExpression(types, path.node.callee)) return

        const remove = createRemove(state)
        const [, arg] = path.node.arguments
        remove(arg)
        if (helpers.isAssignExpression(types, arg)) arg.arguments.forEach(remove)
      },
    },
  }
}
