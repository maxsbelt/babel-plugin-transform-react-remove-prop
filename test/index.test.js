import { transform } from '@babel/core'
import plugin from '../src'
import { DEFAULT_ATTRIBUTE } from '../src/helpers'

const testSnapshot = (code, opts) => {
  expect(
    transform(
      code,
      {
        plugins: [
          '@babel/plugin-syntax-jsx',
          [plugin, opts],
        ],
      },
    ).code,
  ).toMatchSnapshot()
}

describe('babel-plugin-transform-react-remove-prop', () => {
  it('removes attribute', () => testSnapshot(
    `React.createElement("div", { "${DEFAULT_ATTRIBUTE}": "Standard" })`,
  ))

  it('removes attribute if called without React', () => testSnapshot(
    `createElement("div", { "${DEFAULT_ATTRIBUTE}": "Standard" })`,
  ))

  it('works correctly if there is no properties', () => testSnapshot(
    'createElement("div", null, "Children")',
  ))

  it('keeps other attributes', () => testSnapshot(
    `React.createElement("div", { "${DEFAULT_ATTRIBUTE}": "Standard", "test-id": "Standard", label: "Standard" })`,
  ))

  it('keeps `createElement` arguments if it was called on custom object', () => testSnapshot(
    `obj.createElement("div", { "${DEFAULT_ATTRIBUTE}": "custom obj" })`,
  ))

  it('works with Object.assign', () => testSnapshot(
    `React.createElement("div", Object.assign({}, { "${DEFAULT_ATTRIBUTE}": "Custom obj", label: "Custom obj" }, { "another-attr": "Custom obj" }))`,
  ))

  it('works with object spread syntax', () => testSnapshot(
    `React.createElement("div", { ...props, "${DEFAULT_ATTRIBUTE}": "Custom obj", label: "Custom obj" })`,
  ))

  describe('preact', () => {
    it('removes attribute', () => testSnapshot(
      `h("div", { "${DEFAULT_ATTRIBUTE}": "Preact", label: "Preact" })`,
    ))
  })

  describe('JSX syntax', () => {
    it('removes attribute', () => testSnapshot(
      `<div ${DEFAULT_ATTRIBUTE}="JSX" test-id="JSX" label="JSX" />`,
    ))

    it('removes test-id attribute from spread', () => testSnapshot(
      `<div {...{ "${DEFAULT_ATTRIBUTE}": "Spread", label: "Spread" }} />`,
    ))
  })

  describe('`attrs` parameter', () => {
    it('removes necessary attributes', () => testSnapshot(
      'h("div", { "custom-id-1": "custom-1", "custom-id-2": "custom-2", "custom-id-3": "custom-3" })',
      {
        attrs: ['custom-id-1', 'custom-id-2'],
      },
    ))
  })

  describe('non react', () => {
    it('does not remove Identifier parameters', () => testSnapshot(
      `func("div", { "${DEFAULT_ATTRIBUTE}": "Keep it" })`,
    ))

    it('does not remove MemberExpression parameters', () => testSnapshot(
      `some.func("div", { "${DEFAULT_ATTRIBUTE}": "Keep it" })`,
    ))
  })
})
