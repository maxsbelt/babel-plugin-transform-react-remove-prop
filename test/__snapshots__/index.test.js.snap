// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-transform-react-remove-prop \`attrs\` parameter removes necessary attributes 1`] = `
"h(\\"div\\", {
  \\"custom-id-3\\": \\"custom-3\\"
});"
`;

exports[`babel-plugin-transform-react-remove-prop JSX syntax removes attribute 1`] = `"<div test-id=\\"JSX\\" label=\\"JSX\\" />;"`;

exports[`babel-plugin-transform-react-remove-prop JSX syntax removes test-id attribute from spread 1`] = `
"<div {...{
  label: \\"Spread\\"
}} />;"
`;

exports[`babel-plugin-transform-react-remove-prop keeps \`createElement\` arguments if it was called on custom object 1`] = `
"obj.createElement(\\"div\\", {
  \\"data-test-id\\": \\"custom obj\\"
});"
`;

exports[`babel-plugin-transform-react-remove-prop keeps other attributes 1`] = `
"React.createElement(\\"div\\", {
  \\"test-id\\": \\"Standard\\",
  label: \\"Standard\\"
});"
`;

exports[`babel-plugin-transform-react-remove-prop non react does not remove Identifier parameters 1`] = `
"func(\\"div\\", {
  \\"data-test-id\\": \\"Keep it\\"
});"
`;

exports[`babel-plugin-transform-react-remove-prop non react does not remove MemberExpression parameters 1`] = `
"some.func(\\"div\\", {
  \\"data-test-id\\": \\"Keep it\\"
});"
`;

exports[`babel-plugin-transform-react-remove-prop preact removes attribute 1`] = `
"h(\\"div\\", {
  label: \\"Preact\\"
});"
`;

exports[`babel-plugin-transform-react-remove-prop removes attribute 1`] = `"React.createElement(\\"div\\", {});"`;

exports[`babel-plugin-transform-react-remove-prop removes attribute if called without React 1`] = `"createElement(\\"div\\", {});"`;

exports[`babel-plugin-transform-react-remove-prop works correctly if there is no properties 1`] = `"createElement(\\"div\\", null, \\"Children\\");"`;

exports[`babel-plugin-transform-react-remove-prop works with Object.assign 1`] = `
"React.createElement(\\"div\\", Object.assign({}, {
  label: \\"Custom obj\\"
}, {
  \\"another-attr\\": \\"Custom obj\\"
}));"
`;
